airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/report_streams.py

```diff
@@ -0,0 +1,310 @@
+        }
+
+    @abstractmethod
+    def report_init_endpoint(self, record_type: str) -> str:
+        """
+        :return: endpoint to initial report generating process.
+        """
+
+    @property
+    @abstractmethod
+    def metrics_map(self) -> Dict[str, List]:
+        """
+        :return: Map record type to list of available metrics
+        """
+
+    def _check_status(self, report_info: ReportInfo) -> Tuple[Status, str]:
+        """
+        Check report status and return download link if report generated successfuly
+        """
+        check_endpoint = f"/v2/reports/{report_info.report_id}"
+        resp = self._send_http_request(urljoin(self._url, check_endpoint), report_info.profile_id)
+        resp = ReportStatus.parse_raw(resp.text)
+        return resp.status, resp.location
+
+    @backoff.on_exception(
+        backoff.expo,
+        (
+            requests.exceptions.Timeout,
+            requests.exceptions.ConnectionError,
+            TooManyRequests,
+        ),
+        max_tries=5,
+    )
+    def _send_http_request(self, url: str, profile_id: int, json: dict = None):
+        headers = self._get_auth_headers(profile_id)
+        if json:
+            response = self._session.post(
+                url,
+                headers=headers,
+                json=json,
+            )
+        else:
+            response = self._session.get(
+                url,
+                headers=headers,
+            )
+        if response.status_code == HTTPStatus.TOO_MANY_REQUESTS:
+            raise TooManyRequests()
+        return response
+
+    @staticmethod
+    def get_report_date_ranges(start_report_date: datetime) -> List[str]:
+        """
+        Generates dates in YYYYMMDD format for each day started from start_report_date until current date (current date included)
+        """
+        now = datetime.now()
+        if not start_report_date:
+            start_report_date = now
+
+        for days in range(0, (now - start_report_date).days + 1):
+            next_date = start_report_date + timedelta(days=days)
+            next_date = next_date.strftime(ReportStream.REPORT_DATE_FORMAT)
+            yield next_date
+
+    def stream_slices(
+        self, sync_mode: SyncMode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        if sync_mode == SyncMode.full_refresh:
+            start_date = None
+        else:
+            # incremental stream
+            stream_state = stream_state or {}
+            start_date = stream_state.get(self.cursor_field)
+            if start_date:
+                start_date = datetime.strptime(start_date, ReportStream.REPORT_DATE_FORMAT)
+                # We already processed records for date specified in stream state, move to the day after
+                start_date += timedelta(days=1)
+
+        return [{self.cursor_field: date} for date in ReportStream.get_report_date_ranges(start_date)] or [None]
```

Comment:
```suggestion
        return [{self.cursor_field: date} for date in ReportStream.get_report_date_ranges(start_date)]
```
Why can't we do this and delete the 117 and 118 lines of this file?

Reply:
It could return empty  array  (if current state is greater or equal than todays date). In this case read_records wont be called by the abstract source and state don't be updated. That would result in endless loop.